### PR TITLE
feat: remove isFocus from meta state in InputWidgetV2

### LINKED ISSUE #36446: [Feature]: Remove isFocus and isDirty prop from evaluations in InputField
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

I have noticed that for an onFocus event which sets isFocussed meta value to true of an inputField triggers an evaluation. What is the necessity for triggering an evaluation here, I thought another widget could tap this widget's focussed state but that property does not appear to be accessible there. Is it necessary an evaluation needs to be triggered here can we remove it? If there are multiple input widgets in a page i can imagine evaluations to be frequently triggered just by the user moving his cursor around.

### Why should this be worked on?

Reducing the number of evaluations run on a page will greatly improve the performance of the application.

### PR DESCRIPTION
**Pull request by jacquesikot**
## Description

**Problem**
A redundant evaluation cycle is run every time a user focuses on an InputWidget component and the isFocused meta state is updated

**Root Cause**
`isFocus` property of InputWidgetV2 is a meta property and updated through `this.updateWidgetMetaProperty`, thereby changes to these property triggers an evaluation cycle run

**Solution**
Implement the `isFocus` property as a local state of the InputWidgetV2 component

Fixes #36446

## Automation

/ok-to-test tags="@tag.Widget, @tag.Input"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/11362334893>
> Commit: 80e9089681456b7b3547cc4781abf719c561d27d
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=11362334893&attempt=1" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.Widget, @tag.Input`
> Spec:
> <hr>Wed, 16 Oct 2024 10:22:00 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [ ] No

## Summary by CodeRabbit

- **New Features**
	- Enhanced focus management for the InputWidget, allowing for improved internal state handling.
	- Multi-line text input now supports submitting with keyboard shortcuts for a more streamlined user experience.

- **Bug Fixes**
	- Resolved issues with widget metadata by excluding the `isFocused` property from the meta properties.
	- Improved internal state management for input changes, reducing reliance on external properties.

### CODE REVIEW

File: app/client/src/widgets/InputWidgetV2/widget/index.tsx [hunk 1]
```diff
@@ -641,7 +648,10 @@
   // TODO: Fix this the next time the file is edited
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   static getMetaPropertiesMap(): Record<string, any> {
-    return merge(super.getMetaPropertiesMap(), {
+    const baseMetaProperties = super.getMetaPropertiesMap();
+    const rest = omit(baseMetaProperties, ["isFocused"]);
+
+    return merge(rest, {
```
[the remark this answers is not in the document]
Comment (rahulbarwal) [reply app/client/src/widgets/InputWidgetV2/widget/index.tsx:654]: @jacquesikot This makes sense.

File: app/client/src/widgets/InputWidgetV2/widget/index.tsx [hunk 2]
```diff
@@ -641,7 +648,10 @@ class InputWidget extends BaseInputWidget<InputWidgetProps, WidgetState> {
   // TODO: Fix this the next time the file is edited
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   static getMetaPropertiesMap(): Record<string, any> {
-    return merge(super.getMetaPropertiesMap(), {
+    const baseMetaProperties = super.getMetaPropertiesMap();
+    const rest = omit(baseMetaProperties, ["isFocused"]);
```
Comment (rahulbarwal) [inline app/client/src/widgets/InputWidgetV2/widget/index.tsx:652]: Help me understand, why we decided to not completely get rid of `isFocused` from properties and instead using `omit` to remove it?

File: app/client/src/widgets/InputWidgetV2/widget/index.tsx [hunk 3]
```diff
@@ -641,7 +648,10 @@ class InputWidget extends BaseInputWidget<InputWidgetProps, WidgetState> {
   // TODO: Fix this the next time the file is edited
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   static getMetaPropertiesMap(): Record<string, any> {
-    return merge(super.getMetaPropertiesMap(), {
+    const baseMetaProperties = BaseInputWidget.getMetaPropertiesMap();
+    const rest = omit(baseMetaProperties, ["isFocused"]);
```
Comment (rishabhrathod01) [inline app/client/src/widgets/InputWidgetV2/widget/index.tsx:652]: Could we skip using `omit` here by removing `isFocused` from the meta property of the parent widget too?

File: app/client/src/widgets/InputWidgetV2/widget/index.tsx [hunk 4]
```diff
@@ -300,6 +300,13 @@ function InputTypeUpdateHook(
 }
 
 class InputWidget extends BaseInputWidget<InputWidgetProps, WidgetState> {
+  constructor(props: InputWidgetProps) {
+    super(props);
+    this.state = {
+      isFocused: false,
```
Comment (rishabhrathod01) [inline app/client/src/widgets/InputWidgetV2/widget/index.tsx:306]: Github issue also mentions `isDirty` state, did we intentionally skipped that?

Comment (jacquesikot) [reply app/client/src/widgets/InputWidgetV2/widget/index.tsx:306]: Yes, I intentionally skipped that, want to address them one after the other.

File: app/client/src/widgets/InputWidgetV2/widget/index.tsx (hunk 3 above)
Comment (jacquesikot) [reply app/client/src/widgets/InputWidgetV2/widget/index.tsx:652]: Yes, done. I wanted to be sure the base Input Widget isFocus is not implemented anywhere else

File: app/client/src/widgets/InputWidgetV2/widget/index.tsx (hunk 2 above)
Comment (jacquesikot) [reply app/client/src/widgets/InputWidgetV2/widget/index.tsx]: I have gotten rid of it from the base class

**Review by rahulbarwal — APPROVED** [review]

**Pull request merged**